### ISSUE TITLE
More flexible add_node()

### ISSUE THREAD

**Issue by viathor** [google/TensorNetwork#62]
Currently, `add_node()` requires one to specify a tensor. Certain types of nodes are more efficient to manipulate based on their abstract properties than based on their component representation (e.g. copy tensors, Levi-Civita tensors).

We should generalize the API to allow one to add nodes into the network without explicitly computing those tensors. For example, we could re-define `add_node()` to take a Node instead of a Tensor and leave Node construction to the caller. This would allow us to do things such as:

```python
tensor = np.array(...)
net.add_node(tn.Node(tensor))

net.add_node(tn.CopyNode(rank=2, dimension=3))

net.add_node(tn.LeviCivitaNode(rank=4))
```

Of course in practice most nodes are created from multidimensional arrays, so for convenience we could also offer a shortcut e.g. `add_tensor()` such that the following two code snippets are equivalent:

shortcut:

```python
net.add_tensor(tensor)
```

long form:

```python
net.add_node(tn.Node(tensor))
```

**Comment by Thenerdstation** [comment]
I weary of changing the `add_node` API, but I agree adding higher level APIs would be good for common tensors.

Can we do something like `net.add_copy_tensor(rank=2, dimension=3)`?

**Comment by viathor** [comment]
We're already doing it, see #64. Existing `add_node()` introduces strong coupling between TensorNetwork and Node which makes it hard to add new node types.

Also, I'm concerned with cluttering the API with `add_*_node()` methods in TensorNetwork for various node types that we may introduce and with code duplication in those methods.

**Comment by sleichen** [comment]
I think `net.add_node(tn.Node(tensor))` is too verbose. Can we make `net.add_node()' parse its input so that `net.add_node(tensor)' still works and is equivalent to `net.add_node(tn.Node(tensor))`?

**Comment by Thenerdstation** [comment]
Yeah I want gonna suggest that. It would be pretty easy to do as long as all nodes inherit `Node`

**Comment by Thenerdstation** [comment]
If you want to make the tensors of the node lazily evaluated, go ahead and make `node.tensor` a `@property`. I was actually planning on doing this already and deprecate `get_tensor()`
